libcnb/src/layer/public_interface.rs

```diff
@@ -0,0 +1,192 @@
+use crate::build::BuildContext;
+use crate::data::layer_content_metadata::{LayerContentMetadata, LayerTypes};
+use crate::generic::GenericMetadata;
+use crate::layer_env::LayerEnv;
+use crate::Buildpack;
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use std::path::{Path, PathBuf};
+
+/// Represents a buildpack layer written with the libcnb framework.
+///
+/// Buildpack authors implement this trait to define how a layer is created/updated/removed
+/// depending on its state. To use a `Layer` implementation during build, use
+/// [`BuildContext::handle_layer`](crate::build::BuildContext::handle_layer).
+#[allow(unused_variables)]
+pub trait Layer {
+    /// The buildpack this layer is used with.
+    type Buildpack: Buildpack;
+
+    /// The metadata type for this layer. This is the data within `[metadata]` of the layer content
+    /// metadata. If you wish to use raw, untyped, TOML data instead, use [`GenericMetadata`](crate::generic::GenericMetadata).
+    ///
+    /// If the layer metadata cannot be parsed into this type, libcnb will call [`migrate_incompatible_metadata`](Self::migrate_incompatible_metadata)
+    /// with the layer's metadata as raw TOML. This allows migration of older metadata.
+    type Metadata: DeserializeOwned + Serialize + Clone;
+
+    /// Returns the types of this layer. Will be called by libcnb whenever it needs to determine
+    /// the types for this layer. This includes, but is not limited to, after create, update and
+    /// when the layer is not modified at all.
+    ///
+    /// # Implementation Requirements
+    /// Implementations **MUST** be pure. This includes that they **MUST NOT** side-effect,
+    /// including writing to stdout/stderr or the file system.
+    fn types(&self) -> LayerTypes;
+
+    /// Creates the layer from scratch.
+    ///
+    /// `layer_path` will be an empty directory where this method can write files to. Layer
+    /// metadata, including environment variables, is part of the return value of this function and
+    /// will be written to the appropriate locations by libcnb automatically.
+    ///
+    /// # Implementation Requirements
+    /// Implementations **MUST NOT** write to any other location than `layer_path`.
+    fn create(
+        &self,
+        context: &BuildContext<Self::Buildpack>,
+        layer_path: &Path,
+    ) -> Result<LayerResult<Self::Metadata>, <Self::Buildpack as Buildpack>::Error>;
```

Comment:
Could you add an `# Errors` section here and to the other newly added public interfaces?
https://rust-lang.github.io/api-guidelines/documentation.html#function-docs-include-error-panic-and-safety-considerations-c-failure

This will give us less to do for #53.

Reply:
I'm not sure what we're trying to solve here except satisfying the (pedantic) lint. We do not really know anything about the error type in this context as it is defined by the buildpack (`Buildpack::Error`).

Reply:
Ok, let's discuss when we come to enable the lint. I agree a comment is low value here, however I still think there is value in the lint in other cases (we can always `allow` in the low value cases if it comes to it).